library/alloc/src/collections/btree/node.rs

```diff
@@ -1637,15 +1637,17 @@ pub mod marker {
     pub struct ValMut<'a>(PhantomData<&'a mut ()>);
 
     pub trait BorrowType {
-        // Whether node references of this borrow type allow traversing
-        // to other nodes in the tree.
-        const PERMITS_TRAVERSAL: bool = true;
+        // If node references of this borrow type allow traversing to other
+        // nodes in the tree, this constant can be evaluated. Thus reading it
+        // serves as a compile-time assertion.
+        const TRAVERSAL_PERMIT: () = ();
     }
     impl BorrowType for Owned {
-        // Traversal isn't needed, it happens using the result of `borrow_mut`.
+        // Reject evaluation, because traversal isn't needed. Instead traversal
+        // happens using the result of `borrow_mut`.
         // By disabling traversal, and only creating new references to roots,
         // we know that every reference of the `Owned` type is to a root node.
-        const PERMITS_TRAVERSAL: bool = false;
+        const TRAVERSAL_PERMIT: () = panic!();
     }
```

Comment:
Can you add some UI tests to show the cases where this does and doesn't panic and to let us review the panic output?

Reply:
I don't think I can. These static asserts guard the private internals of node.rs. We have unit test of those internals, we can do positive compile tests, but a negative compile test of internals, I doubt the test framework caters for that. And surely we don't want to make this stuff public.

The way I tested this is to try to address the remaining FIXME on `reborrow_mut`in node.rs: disable the permit for `Mut<'+>` like it is disabled for `Owned`, and build errors point out the first place where a `Mut` handle is traversed. While without this PR, you have to run tests to see it, and you have to tests to offer complete coverage. I've edited the PR description a bit.